### PR TITLE
Change add-on to use inputstream.ffmpegdirect which also enables timeshifting

### PR DESCRIPTION
**Pull request by phunkyfish**
v21.0.0
- Change add-on to use inputstream.ffmpegdirect which also enables timeshifting
- Fix jsoncpp deprecation warnings
- Add build script for mac

### CODE REVIEW

**Comment by phunkyfish** [comment]
Note this PR requires testing before any merge is possible.

File: pvr.octonet/resources/settings.xml
Comment (ksooo) [inline pvr.octonet/resources/settings.xml]: Typo. Lodi -> Kodi

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: ???? XXXXX

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: ??? XXXXX

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: typo: inpustream -> inputstream

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: typo: inpustream -> inputstream

File: pvr.octonet/addon.xml.in
Comment (ksooo) [inline pvr.octonet/addon.xml.in]: 21.0.0 as this is the first Omega version

File: pvr.octonet/changelog.txt
Comment (ksooo) [inline pvr.octonet/changelog.txt]: 21.0.0

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (ksooo) [inline pvr.octonet/resources/language/resource.language.en_gb/strings.po]: Will anybody understand what this means? Can we find better wording?

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (ksooo) [inline pvr.octonet/resources/language/resource.language.en_gb/strings.po]: What is live stream api? Can we find better wording?

**Review by ksooo — CHANGES_REQUESTED** [review]
Needs a bit more love, imo.

Q: could we get rid of the internal rtsp stuff completely and require inputstream.adaptive?
Q2: why so many different stream handlers? Wouldn't be one (ISA?) enough?

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (da-anda) [inline pvr.octonet/resources/language/resource.language.en_gb/strings.po]: maybe change this to `The IP/address of the octonet server. Note that in case of an address, the protocol (http://) has to be omitted.`

File: src/OctonetData.cpp
Comment (phunkyfish) [reply src/OctonetData.cpp]: Done

File: pvr.octonet/resources/settings.xml
Comment (phunkyfish) [reply pvr.octonet/resources/settings.xml]: Done

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (phunkyfish) [reply pvr.octonet/resources/language/resource.language.en_gb/strings.po]: Done

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (phunkyfish) [reply pvr.octonet/resources/language/resource.language.en_gb/strings.po]: Done

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (phunkyfish) [reply pvr.octonet/resources/language/resource.language.en_gb/strings.po]: Done

File: src/OctonetData.cpp
Comment (phunkyfish) [reply src/OctonetData.cpp]: Done, fixed with proper log statement

File: src/OctonetData.cpp
Comment (phunkyfish) [reply src/OctonetData.cpp]: Done

File: pvr.octonet/addon.xml.in
Comment (phunkyfish) [reply pvr.octonet/addon.xml.in]: Done

File: src/OctonetData.cpp
Comment (phunkyfish) [reply src/OctonetData.cpp]: Done

File: pvr.octonet/changelog.txt
Comment (phunkyfish) [reply pvr.octonet/changelog.txt]: Done

**Comment by phunkyfish** [comment]
I've cleaned up the PR now. Extra log statements were simply for testing.

**Comment by phunkyfish** [comment]
> Needs a bit more love, imo.
> 
> Q: could we get rid of the internal rtsp stuff completely and require inputstream.adaptive? Q2: why so many different stream handlers? Wouldn't be one (ISA?) enough?

Yes, getting rid of the internal rtsp stuff is the plan. But i can't test here, need @da-anda to test and see what works. ISA can't be used as it does not support rtsp.

**Comment by phunkyfish** [comment]
@da-anda I also added a commit to request kodi update the EPG if it has not been requested. Should mean you always get EPG now.

**Comment by ksooo** [comment]
> @da-anda I also added a commit to request kodi update the EPG if it has not been requested. Should mean you always get EPG now.

not sure this is a good idea as fetching the epg takes a very long time on the server side and what happens if you have multiple pvr.octonet connected to the same device all doing this (regularly)? I think this problem needs to be solved in the firmware of the octonet. There a feature is missing that you can set a fetch interval for the epg data so that the octo takes care. 

EDIT: oh i misread your comment, ignore what i wrote

File: pvr.octonet/resources/language/resource.language.en_gb/strings.po
Comment (ksooo) [inline pvr.octonet/resources/language/resource.language.en_gb/strings.po]: Typo: inpustream.rtmp -> inputstream.rtmp

**Comment by da-anda** [comment]
I tested all 3 alternatives, and only `KODI inputstream` and `ffmpegdirect` ended up with a good picture quality. RTMP worked, but had the same issues as the current implementation. Channel switch times were slightly slower though, from what I can tell

File: pvr.octonet/resources/settings.xml
Comment (ksooo) [inline pvr.octonet/resources/settings.xml]: You should adjust the comments to match the translatable strings.

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: This should be removed as we are now using the thread.

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: This sentence somehow reads wrong ... then request then ask ...

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: Q: Have we ever encountered this during our tests?

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: typo inpustream

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: If you end up here, is this an error condition (return PVR_ERROR_...)?
At least the following log will have an empty `openUsingDescription`, which seems not intended.

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: typo supoprt

File: src/OctonetData.cpp
Comment (ksooo) [inline src/OctonetData.cpp]: typo inpustream

**Review by ksooo — CHANGES_REQUESTED** [review]

**Comment by ksooo** [comment]
Please fix the commit message "changelog and version v21.1.0" - version is 21.0.0

File: src/OctonetData.cpp
Comment (da-anda) [reply src/OctonetData.cpp]: you mean EPG not being loaded on startup? Yes. I had to clear PVR data in order to see the HTTP request for the EPG in the logs. So despite not returning EPG for the current selection of unicast channels (which seems to be working now on the octonet) the add-on (or KODI) didn't even try to fetch it most of the time

**Comment by phunkyfish** [comment]
> Please fix the commit message "changelog and version v21.1.0" - version is 21.0.0

Done

File: src/OctonetData.h
Comment (da-anda) [inline src/OctonetData.h]: I suppose this enum is no longer needed

**Comment by phunkyfish** [comment]
Ok, I removed the original setting, and opening the stream using the Live API. Now we just use ffmpegdirect, which supports timeshifting.

Note, a fix is required for timeshifting for rtsp, but that is a separate PR for ffmpegdirect.

File: src/OctonetData.cpp
Comment (ksooo) [reply src/OctonetData.cpp]: > the add-on (or KODI) didn't even try to fetch it most of the time

IIRC kodi remembers the time stamp the EPG was obtained last time successfully and fetches EPG next time the EPG fetch interval is over (it's a setting).

Could it be (need to check) that even if the octonet does not return any EPG data (for whatever reason), pvr.octonet returns success to Kodi?

Comment (ksooo) [reply src/OctonetData.cpp]: https://github.com/DigitalDevices/pvr.octonet/blob/Omega/src/OctonetData.cpp#L354

Yes, GetEPGForChannel always returns "success".

Comment (da-anda) [reply src/OctonetData.cpp]: in my case it returned data, but the data could not be mapped to any channel. So while it had data, it also didn't in a way. No idea which state got reported back to KODI

**Comment by ksooo** [comment]
> Now we just use ffmpegdirect, which supports timeshifting.

Cool, could we add a dependency to ffmpegdirect to addon.xml?

**Comment by ksooo** [comment]
I'm not conviced that the EPG load 10 secs after start is something that actually fixes the root cause of the "@da-anda not having EPG data" problem. If there are no data at the octonet, you can try to fetch them as many times as you want without success. Kodi remembering last EPG fetch timestamp and da-anda meanwhile fixing EPG data on the octopus and Kodi not fetching these new data on next start is something that the commit in this PR fixes by "accident" I'd say.

**Comment by da-anda** [comment]
> Cool, could we add a dependency to ffmpegdirect to addon.xml?

it already is added, isn't it?

File: pvr.octonet/addon.xml.in
Comment (da-anda) [inline pvr.octonet/addon.xml.in:8]: @ksooo here is the dependency - or are we missing something else?

File: src/OctonetData.cpp
Comment (ksooo) [reply src/OctonetData.cpp]: > No idea which state got reported back to KODI

Always "success". GetEPGForChannel never reports any error. For me the question remains, what the actual root cause for the bug is. The load after ten secs is just a golden hammer, not a real fix. We did not even understand what actually happened.

**Comment by da-anda** [comment]
> I'm not conviced that the EPG load 10 secs after start is something that actually fixes the root cause of the "@da-anda not having EPG data" problem. If there are no data at the octonet, you can try to fetch them as many times as you want without success. Kodi remembering last EPG fetch timestamp and da-anda meanwhile fixing EPG data on the octopus and Kodi not fetching these new data on next start is something that the commit in this PR fixes by "accident" I'd say.

according to your other comment, the correct fix would be to not return the NO_ERROR state in case no channel could be mapped, correct? (aka some data got back from octonet, but not the correct channels)

File: pvr.octonet/addon.xml.in
Comment (ksooo) [reply pvr.octonet/addon.xml.in:8]: No.

**Comment by ksooo** [comment]
> according to your other comment, the correct fix would be to not return the NO_ERROR state in case no channel could be mapped, correct?

Not sure. It should return NO_ERROR only if data for the requested channel was returned by octopus, but for channels not having EPG data also NO_ERROR must be returned (afair, it's been a while since I looked at this stuff). Not sure whether the octopus API allows to detect the latter.

File: src/OctonetData.h
Comment (phunkyfish) [reply src/OctonetData.h]: Removed

**Comment by phunkyfish** [comment]
> > according to your other comment, the correct fix would be to not return the NO_ERROR state in case no channel could be mapped, correct?
> 
> Not sure. It should return NO_ERROR only if data for the requested channel was returned by octopus, but for channels not having EPG data also NO_ERROR must be returned (afair, it's been a while since I looked at this stuff). Not sure whether the octopus API allows to detect the latter.

So this behaviour was observed when @da-anda did not have a valid EPG. Now that he does I reckon we drop this commit for now and see if it's needed later at a later point.

In recent tests Kodi remembers the EPG data, and should just request it according to it's own schedule.

File: src/OctonetData.cpp
Comment (phunkyfish) [reply src/OctonetData.cpp]: I've removed this commit now.

**Comment by phunkyfish** [comment]
Ok, commit dropped, it's in draft PR #68 in case we need it.

This PR should be ready to go now.

**Review by ksooo — APPROVED** [review]
Awesome!

**Comment by ksooo** [comment]
@phunkyfish did a runtime-test of this PR. Timeshift seems not to work properly. 

Sometimes Kodi completely freezes when trying ff/rew, sometimes Kodi crashes. There seem to be several scenarios to cause misbehaviour.

**Comment by phunkyfish** [comment]
> @phunkyfish did a runtime-test of this PR. Timeshift seems not to work properly.
> 
> Sometimes Kodi completely freezes when trying ff/rew, sometimes Kodi crashes. There seem to be several scenarios to cause misbehaviour.

Correct, as stated above, this needs to be fixed in a separate PR for ffmpegdirect. But nothing it change in this PR to help that.

**Comment by phunkyfish** [comment]
> > @phunkyfish did a runtime-test of this PR. Timeshift seems not to work properly.
> > Sometimes Kodi completely freezes when trying ff/rew, sometimes Kodi crashes. There seem to be several scenarios to cause misbehaviour.
> 
> Correct, as stated above, this needs to be fixed in a separate PR for ffmpegdirect. But nothing it change in this PR to help that.

If you could do a backtrace and pinpoint the crash and logs too that would be great. When you have time.

**Comment by ksooo** [comment]
> Correct, as stated above, this needs to be fixed in a separate PR for ffmpegdirect. But nothing it change in this PR to help that.

Okay, then let’s hold off the merge of this PR until the ffmpegdirect fix is available.

**Comment by phunkyfish** [comment]
> > Correct, as stated above, this needs to be fixed in a separate PR for ffmpegdirect. But nothing it change in this PR to help that.
> 
> Okay, then let’s hold off the merge of this PR until the ffmpegdirect fix is available.

This would depend on me having access to a stream. Which I don't currently. I could just comment out the setting for now and then we can revert that commit once it's working?

**Comment by da-anda** [comment]
well, technically you do have access to a stream since VLC can play the streams off of my octonet just fine, it's just KODI/ffmpeg that won't play it via remote connection for whatever reason 😉

**Comment by phunkyfish** [comment]
Well it turns out that timeshift is broken on omega. So hopefully I can fix this without a stream in rtsp stream kodi. Time to start bisecting…

**Comment by ksooo** [comment]
> Well it turns out that timeshift is broken on omega.

Okay, then let's merge this PR as is.

**Pull request merged**
